### PR TITLE
[CDAP-2363] Provide a reason for adapter deletion failure. Also change t...

### PR DESCRIPTION
**Pull request by bdmogal**
...he response code to CONFLICT from FORBIDDEN

Jira: [CDAP-2363](https://issues.cask.co/browse/CDAP-2363)
Build: http://builds.cask.co/browse/CDAP-RBT259

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java [hunk 1]
```diff
@@ -276,7 +276,7 @@ public void deleteAdapter(HttpRequest request, HttpResponder responder,
       adapterService.removeAdapter(Id.Namespace.from(namespaceId), adapterName);
       responder.sendStatus(HttpResponseStatus.OK);
     } catch (CannotBeDeletedException e) {
-      responder.sendString(HttpResponseStatus.FORBIDDEN, e.getMessage());
+      responder.sendString(HttpResponseStatus.CONFLICT, e.getMessage());
```
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java]: Are most `CannotBeDeletedException` supposed to result in CONFLICT? If so, you can remove this catch-block and simply catch it in https://github.com/caskdata/cdap/blob/develop/cdap-common/src/main/java/co/cask/cdap/common/exception/HttpExceptionHandler.java#L35

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java [hunk 2]
```diff
@@ -276,7 +276,7 @@ public void deleteAdapter(HttpRequest request, HttpResponder responder,
...
     } catch (NotFoundException e) {
```
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java]: Not part of your PR, but you can remove this since `HttpExceptionHandler` handles this. Possibly also the `catch (Throwable t)` block below too.

Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java]: We should probably do a walk through all the handlers and remove unnecessary catch clauses. I created https://issues.cask.co/browse/CDAP-2422. @jwang47 can you add some more details to that, please?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java]: This is what I see so far:

```
$ gj "catch.*\\(CannotBeDeletedException.*\)" -A2 *
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java:    } catch (CannotBeDeletedException e) {
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java-      responder.sendString(HttpResponseStatus.CONFLICT, e.getMessage());
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java-    } catch (NotFoundException e) {
--
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java:      } catch (CannotBeDeletedException cbde) {
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java-        appStatus = AppFabricServiceStatus.PROGRAM_STILL_RUNNING;
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java-      } catch (Exception e) {
--
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java:    } catch (CannotBeDeletedException cbde) {
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java-      status = AppFabricServiceStatus.PROGRAM_STILL_RUNNING;
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java-    } catch (Exception e) {
--
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java:    } catch (NamespaceCannotBeDeletedException e) {
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java-      responder.sendString(HttpResponseStatus.CONFLICT, e.getMessage());
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java-    } catch (Exception e) {
--
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java:    } catch (NamespaceCannotBeDeletedException e) {
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java-      responder.sendString(HttpResponseStatus.FORBIDDEN, e.getMessage());
cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java-    } catch (Exception e) {
```

So its a mix of ``FORBIDDEN`` (delete app, delete namespaces - which has one ``FORBIDDEN`` and another ``CONFLICT``) and ``CONFLICT`` currently. If we were to change everything to ``CONFLICT``, we'd have two API changes. I am fine with standardizing all of these in this PR.

Thoughts?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java (hunk 2 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java]: I'm ok with changing this too - about 7 occurrences of ``NotFoundException`` in ``AdapterHttpHandler``.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java]: Ok, it is too late in the release cycle to change everything right now, I'll add some comments to @anew's jira below and we can do all of these changes together at once during the next release.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java (hunk 2 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java]: @jwang47 let's do this in a separate PR. Tried it, its increasing this PR's scope, not sure I want to do that late in the cycle. Let's do this as part of a bigger cleanup.

**Comment by sreevatsanraman** [comment]
LGTM :+1:  Unrelated failure in build which is already fixed in release.
